### ISSUE TITLE
Rolling out nullness annotations

### ISSUE THREAD

**Issue by kerams** [fable-compiler/Fable#3887]
.NET 9 is around the corner, and with it support for consuming and emitting nullness metadata in F# projects. JS APIs will often return `null`/`undefined`, and while Fable bindings often capture this information by using options, [sometimes we give no indication that `null`s are a possibility](https://github.com/fable-compiler/fable-browser/issues/99). I think F# 9 will be really useful here, so I'd like to start a discussion about the adoption plan for these annotations.

From the technical point of view, binding projects need to be compiled using .NET 9 SDK, `<LangVersion>preview</LangVersion>` (for now) and with the `--checknulls+` flag (also for now - I think this will be aligned with C#'s `<Nullable>enable</Nullable>`). TFMs don't matter, so `netstandard2.0` assemblies will contain nullness metadata too.

Finally, every API will need to be manually reviewed and annotated as required. In the case of existing bindings that use options with reference types, there's the question of whether we want to preserve options or replace them with `null` annotations, just so that a single, unified approach is used.

**Comment by MangelMaxime** [comment]
I decided to move this issue into Fable repository because there is more to decide that the aesthetic of the API.

@ncave As you probably know more about FCS + Fable internals, what is your vision on how nullness could be added to Fable?

**Comment by MangelMaxime** [comment]
> but then we'll also need to add `net9.0` TFMs so that nullness warnings show up in the IDE, and users also need to retarget their Fable projects to `net9.0` so that nullness is checked during compilation (once the Fable compiler is updated).

> I wonder why for this new feature people need to use target `net9.0`?
> 
> ~I mean in previous addition to F# people where able to keep using `netstandard2.1` for example without issues.~
> 
> After thinking, I think I understand why you want to do that. This is to have support for compiler directives to branch out the API with or without nullness.

To avoid the need for targeting `net9.0`, we could also decide to add a custom compiler directive in Fable like `FABLE_NULLNESS`. And we could make it active by default or not starting a certain version of Fable.

**Comment by kerams** [comment]
> If I remember how TFM works, in order to use net9.0 in your project you need to use .NET 9.0 SDK

You need .NET 9 to use nullness in the first place, so I don't see why requiring `net9.0` to make use of nullness-aware Fable bindings is a problem.

**Comment by MangelMaxime** [comment]
> > Duplicating the API surface means depending on your configuration you have a different API
> 
> Technically speaking, I'd say it's the same API surface with extra metadata. In any case, that's just how C# and now F# work. With different values of `<Nullable>enable</Nullable>` you get a "different" API surface for methods across the BCL, and if you also use warnings as errors, code that no longer compiles. In my opinion this is just par for the course.
> 
> >

Perhaps, but it still means that as a maintainer you need to double the amount of code you maintains. And personally, I don't think I would do it, having bindings with thousands of lines is already not fun to maintain 😅 (so if I need to double the amount ...)

Especially, because using `Option` has worked good enough for such scenarios.

> > type Nullable<'T> =
> 
> This will give you truly different API surfaces. In Fable JS there isn't going to be any difference in the transpiled output as far as I am aware (options only exist at compile time), but with projects compiled with .NET, you get incompatible IL and run-time differences. Most importantly, you will also need different method bodies.

I see, my vision was primarily focused on Fable because the original discussion was about bindings. And from Fable POV, `option` or `| null` will probably be the same. So `Nullable` was only aimed to be used for interop not libraries.

Then perhaps the way to mimic what you are describing is:

```fs
type Nullable<'T> =
    #if NET9_0_OR_GREATER
    'T | null
    #else
    'T
    #endif
```

For me, it is also difficult to take a decision on such things without being able to play with the code. Because, there are a lot of things that I don't see without having a compiler result / code output to check.

**Comment by T-Gro** [comment]
As an independent observer, I would not recommend replacing existing options with `T | null`. The option type has stronger support both in the Fsharp.Core standard library as well as in the surrounding ecosystem, it is a more idiomatic container to use.

I would consider adding the Nullable annotation around values that have not been optional until now, but were practically shown to be null at runtime in various cases.
To enable conditional compilation, the `<Nullable>`project property is adding an  automatic `--define:NULLABLE`. Which could then be used to decide between `'T | null` and `'T`

**Comment by ncave** [comment]
@MangelMaxime Currently Fable transpiles F# options to `undefined` (when `None`), I wonder if that would be a better default for `Nullable<T>` as well. But I don't maintain a lot of JS bindings, nor do I have strong opinions about it, so whatever works best, I just agree with others that we should try to keep using options where we can. Perhaps reasoning about an actual JS binding (or a fragment of) as an example can help clarify what works well.

On a side note, the best explanation of [nullable types](https://dart.dev/null-safety/understanding-null-safety#non-nullable-and-nullable-types) I've ever read was, oddly enough, in Dart documentation.

**Comment by MangelMaxime** [comment]
Thanks both of you for your input.

Personally, my current position is:

1. We add support for `nullness` in Fable, because we want feature parity regarding F# specs/language.

	I think that the initial implementation should probably behave like `option`. If there is a value output that value otherwise generate
	
	- `null` for Python, Rust, Dart, etc. 
	- `undefined` for JavaScript? Or do we want `null`, allowing Fable to have `None` to represents `undefined` and `| null` for `null`.
	
	I think `undefined` is less confusing. If a user want to distinguish between `T`, `null`, `undefined` they are doing something really special and could use a specialised type perhaps?

2. We keep recommending `option` in bindings because up to now it worked fine. Thanks to that, it give time for every one to experiment with the new `nullness` support.

	For example, F# community need to figure out its usage in normal F# code regarding the existing `option` type. @T-Gro seems to say that `Option` has a stronger support so perhaps people will limit `nullness` for a safer usage of BCL/C# code.
	
	From Fable point of view, if `nullness` is deemed a better candidate/experimentation for binding (both for .NET and others runtimes). It will give us the time, to check what impact it has. 
	
	I am saying a better candidate here, because if the changes are just a "rename" then I think keeping `option` is better because it will give a single way of representing `| null` or `| undefined` from TypeScript, etc. With the benefit of not forking the ecosystem. We need to consider that now days, existing libraries for Fable evolve slowly compare to before. Look at the Fable.React 9.0.0 + Feliz 2 delays for upgrading the libraries (some of them required a fork..)
	
	This is both because:
	
	- In general once a library is produced using F# then it tends to just works
	- And, because we have a fewer number of maintainers

3. We also need to think how does `nullness` works in regard to libraries that works both for .NET + Fable runtimes.

	- Does it works out of the box?
	- Do we need to create a `type Nullable` and have some Fable internal/replacement magic happening/

	```fs
	type Nullable<'T> =
	    #if NULLABLE
	    'T | null
	    #else
	    'T
	    #endif
	```

	Will this code mimic the behaviour on .NET compared to `string | null` (vs `Nullable<string>`). How does it behave for Fable, and what does it output?

**Comment by MangelMaxime** [comment]
I am starting to look at how to implement `Nullable` in Fable.

I think we could map `Nullable` to check against only `null` in JavaScript. My idea being, once F# supports anonymous DUs which is planned and is waiting for the C# implementation of it.

We could make some Fable magic to map `| undefined` to check against `undefined`.

This would gives us the ability to support `string | null | undefined` just like in TypeScript.

If we map `| null` to check for both `null` and `undefined` like I originally proposed above we will loose the possibility of doing something like `string | null | undefined` without introducing a subtle breaking chance.

**Comment by joprice** [comment]
For comparison, Scalajs did a similar solution https://www.scala-js.org/api/scalajs-library/latest/scala/scalajs/js/index.html#UndefOr[+A]=A@scala.annotation.unchecked.uncheckedVariance|Unit. Scala 3 has union types, but Scalajs supports scala 2 as well, so has it's own union type that uses implicits, similar to `U2` https://www.scala-js.org/api/scalajs-library/latest/scala/scalajs/js/$bar.html

**Comment by ncave** [comment]
We need a way to model certain JS APIs (e.g. `Map.get` or `Array.find`) that return `T | undefined`.
Here are the existing types we have:
```ts
export type Nullable<T> = T | null | undefined;
export type Option<T> = T | Some<T> | undefined;
```
IMO it makes most sense to model the new nullable types to `T | undefined`, instead of one of the existing types. But I could be wrong, so feel free to disagree.

**Comment by MangelMaxime** [comment]
Here are some examples code with their compilation equivalent trying to explains my point of view:

## Mapping nullness to `T | null` in JS/TS

```fs
type NullableInt = int | null;
// TS => type NullableInt = number | null

// Defined in Fable.Core
type undefined = obj; // perhaps obj is not the correct representation but this is for example

type NullableOrUndefinedInt = int | null | undefined;
// TS => type NullableOrUndefinedInt = number | null | undefined
```

## Mapping nullness to `T | undefined` in JS/TS

```fs
type NullableInt = int | null;
// TS => type NullableInt = number | undefined
// Here we have an issue because we can't express the `null` constraint in TS/JS from the F# code

// Unless we defined some custom types like
type jsUndefined = obj
type jsNull = obj

type NullableInt = int | jsNull;
// TS => type NullableInt = number | null

type NullableOrUndefinedInt = int | jsNull | jsUndefined;
// TS => type NullableOrUndefinedInt = number | null | undefined
```

For this reason, I think mapping `Nullness` to `null` offer a more straightforward mapping to JS/TS and allows us to support `undefined` transparently in the future.

**Comment by kerams** [comment]
You're all talking about TS type annotations, right? When targetting JS, I'd expect `string | null` to cover both `null` and `undefined`, because I consider the distinction between the two a language implementation detail that I don't care about as an F# developer outside of low-level interop scenarios.

**Comment by MangelMaxime** [comment]
> You're all talking about TS type annotations, right? 

We are talking about which behavior it should have when targeting JS/TS.

The distinction can be/is needed for interop with JavaScript libraries, in JavaScript `null` and `undefined` are not exactly equivalent.

- `undefined`, means a variable has been declared but not assigned
- `null`, means a variable has been declared and assigned the value `null`

**Comment by kerams** [comment]
OK, but if you're saying that we need to be able to tell the difference in F# too and an API accepts or can return `undefined`, we cannot use nullable reference types, since `string | null | undefinedType` would require https://github.com/fsharp/fslang-suggestions/issues/538

**Comment by MangelMaxime** [comment]
Yes, I know that to do we need `string | null | undefinedType` https://github.com/fsharp/fslang-suggestions/issues/538.

But IHMO we need to plan for when it will be available, anonymous DUs is planned by the F# team but they are waiting for C# to implement it first for compatibility. So "we know" it is coming.

However, if we originally map nullness to `null | undefined` we will not be able to later say that actually nullness map to `null` only as it would introduce a breaking changes that can't be easily spotted, and could break libraries that has been published since then.

**Comment by kerams** [comment]
It's coming who knows when. Regardless, we can close the issue then?

**Comment by MangelMaxime** [comment]
> Reason is, besides not having to wait for anonymous unions, this would also free Fable from .NET-imposed limitations on the F#9 ` | null` feature, mainly around generic code and ability to work with value types.

I don't know well `| null`, so your comment made me read again the documentation of F# nullable.

I didn't find information about "limitations [...] mainly around generic code" but I found this section:

![CleanShot 2025-01-09 at 09 52 27@2x](https://github.com/user-attachments/assets/c81a0a5b-fb7f-4159-9936-78b8f19f2030)

Which seems to defeat my idea of using `| null` to represent `null` in JavaScript, because we can't assign `null` in F# to nullness. I hope my sentence is clear.

Because of that it seems like we need to consider nullness supports as a F# feature, and not so much as a JS/TS equivalent to `null` value (for interop).

I am also not sure what you mean by:

> and ability to work with value types

> For the low level scenarios that need to see a difference (i.e., should not be the default because of ergonomy), is there a reason why something like:
> `type JsLowLevel<'T> = Data of 'T | Undefined | Null`
> would not work?

Here you are using a real DUs right? I suppose we could indeed make it work, I am not sure how much magic from Fable we would need to do it.

**Comment by ncave** [comment]
@MangelMaxime Do you have an example of why can't we just transpile the F# nullable types (`T |  null`) as the wider Fable TS type `Nullable<T>` (e.g. `T | null | undefined`) ?

**Comment by MangelMaxime** [comment]
I don't have one under the hand not, and the more the discussion goes and the more compiling to the existing seems to be ok for me.

My initial reason was to make a distinction in the future and have almost a 1 to 1 match with TypeScript syntax which is always good for bindings but this is not possible.

**Comment by MangelMaxime** [comment]
I think the next step now is to give it a try to compile to:

```ts
export type Nullable<T> = T | null | undefined;
```

add some tests to check how it plays and adapt if needed.

**Comment by T-Gro** [comment]
> Here you are using a real DUs right? I suppose we could indeed make it work, I am not sure how much magic from Fable we would need to do it.

A real DU, yes.

I think you got it correctly: 
The F# 9 ` T | null` construct enforces constraints on T in typechecking: The underlying `T` must be non nullable, and must be a reference type. Meaning you cannot do ` int | null` in F# 9. And if you use ` | null` in a generic wrapper, the type argument constraint is propagated to that wrapper as well.

**Comment by ncave** [comment]
@MangelMaxime 

Then there is the question of how to properly represent the [Nullable Reference Types](https://devblogs.microsoft.com/dotnet/nullable-reference-types-in-fsharp-9/) in Fable.AST.
The best I could come up with (without making large changes to the Fable.AST) was to [wrap them in a Nullable<T>](https://github.com/fable-compiler/Fable/blob/main/src/Fable.Transforms/FSharp2Fable.Util.fs#L1635-L1639) as a marker (not enabled yet).

Once we come up with some tests for `<Nullable>enable</Nullable>`, we can start working on proper support.

**Comment by MangelMaxime** [comment]
You are right when working on #4070, I felt like it was strange we didn't modify the AST.

I suppose the standard way would be to add a new `| Nullable of Type` case to 

https://github.com/fable-compiler/Fable/blob/048f27ec87818ca55bdf441dda0178e55d0b7030/src/Fable.AST/Fable.fs#L184-L202

However, it would make a change to Fable.AST like you said because we would be forced to "unwrap" this case. But it would also help us identify all the places where this `Nullable` information could matter.

There don't seems to be that many places impacted.

![Image](https://github.com/user-attachments/assets/e55831d5-2a58-42e4-9188-b4e17fdec753)

Another impact of doing it this way are Fable plugins, I am not sure if the changes will be transparent or if the 2 main libraries using Fable plugin will need a new release or not.

In fact, I started to start enabling `<Nullable>` to a real world project and because Fable compile using source files, some (most?) libraries are going to need to be re-release with the new nullness information.

https://github.com/fable-compiler/fable-react/commit/49dcb61b3808ad80943aa59bc2a6aab78c8cf946#diff-daab6b52fc16af3b4be6146fb0d78c9756d5443a8364d6c45b44d524fe992641

I will try to add a new case to Fable.AST and test out the behavior regarding Fable plugins. To see how much impact it will have.

------------

> Once we come up with some tests for `<Nullable>enable</Nullable>`, we can start working on proper support.

I did write some "nullness" tests [here](https://github.com/fable-compiler/Fable/blob/048f27ec87818ca55bdf441dda0178e55d0b7030/tests/Js/Main/Nullness.fs). I am not sure if this covers enough or not, as I am still trying to grasp what "nullness" does.

From the look of it, it feels like this is mostly a compiler information and we didn't have too much to do to make the basics works (unless it works by pure luck because I am not testing the correct things 😅).

One of the area that I didn't check, is regarding reflection. I don't know if this should be impacted or if F# introduced a new reflection API to check if a type supports `NullAnnotation`.
